docs/extending/code-overview/glossary.rst

```diff
@@ -175,6 +174,18 @@ Glossary
 
      For the general concept, see `Rewriting <https://en.wikipedia.org/wiki/Rewriting>`_.
 
+   S-expression
+
+     A structure which consists of a sequence atoms, and (nested)
+     expressions. However at each level there is a ``Head`` which
+     represents an operator or function.
```

Comment:
In ``S-Expressions`` the head is always an ``Atom``. But in WL, the head sometimes is another non-atomic expression, like in ``Derivative[1,2][f]``

Reply:
Ok. I just checked in the code, and yes, you are correct.  In rewrite_apply_eval_step `self._head` is:

```
<Expression: System`Derivative[1, 2]>
```

and that is why [line 923](https://github.com/Mathics3/mathics-core/blob/master/mathics/core/expression.py#L923) has the `.evaluate(evaluation)` call. 

Still I am not convinced using M-expression is the right way to go here rather than saying that this is really a _variant_ of an S-expression called an "M-expression" a few places in the docs.  

And the fact that we aren't turning this into an S-expression may also be a source of inefficiency. But as long as line 923 is there, we should make note that this is there because we have such kinds of situations and that this is where the distinction between S-expression and M-expression lies.

Reply:
@mmatera In 6c3469e we have gone back to M-expression, while trying to make it clear that this is more or less the same thing as an S-expression so that people who are familiar with how things normally work, won't get confused by the slight variation here. 

Also in the last comment, I start to define `Symbol`. I suspect that "binding" will ultimately get folded into "Definition", but that's for a different time and PR. 

Lastly, I find it amusing that you are incapable of giving the _simplest _example  (e.g. `Derivative[1][f]`) that gets across the point. I would have thought in the education business this would be a useful thing to be able to be able do.

Reply:
@rocky, I had chosen carefully the "simplest" example that I imagined would avoid thinking that the expression would be equivalent to a single symbol "Derivative$1$f". In any case, it is clear that I failed :(.

Reply:
No problem - please keep it in mind though. Practice makes perfect. And you may be able to use this in the classroom as well. In my opinion it was one of things that made Feynman a great lecturer.

And also, I am a very simple-minded person. I get confused with unnecessary distraction.